patches/server/0686-Missing-Entity-Behavior-API.patch

```diff
@@ -106,6 +106,47 @@ index a4f909123de26d911aea7cd767d2315ed1f697c9..0eee53c068bca070a86645d0ba54fb1a
 +    }
 +    // Paper End - More cat api
  }
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftDolphin.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftDolphin.java
+index 938e141f161acf5de5d3361382b514caea02c6fb..46eb3523c88131ac52f30b5ce008f8a120db7727 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftDolphin.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftDolphin.java
+@@ -24,4 +24,36 @@ public class CraftDolphin extends CraftWaterMob implements Dolphin {
+     public EntityType getType() {
+         return EntityType.DOLPHIN;
+     }
++
++    // Paper start - Dolphin API
++    @Override
++    public int getMoistness() {
++        return this.getHandle().getMoistnessLevel();
++    }
++
++    @Override
++    public void setMoistness(int moistness) {
++        this.getHandle().setMoisntessLevel(moistness);
++    }
++
++    @Override
++    public void setHasFish(boolean hasFish) {
++        this.getHandle().setGotFish(hasFish);
++    }
++
++    @Override
++    public boolean hasFish() {
++        return this.getHandle().gotFish();
++    }
++
++    @Override
++    public org.bukkit.Location getTreasureLocation() {
++        return net.minecraft.server.MCUtil.toLocation(this.getHandle().level, this.getHandle().getTreasurePos());
```

Comment:
Not sure about this... As ugly as the bukkit Location class is, this might have to be a null world. The entity doesn't track the world, its only a BlockPos. adding a world here, might incorrectly have plugins make that assumption that the treasure location the dolphin was actually referring to was in the world associated with the returned location.

Reply:
Although I agree, I took reference from turtle api here. Is it worth the inconsistency? The explanation makes sense but idk, does it matter?

Reply:
![image](https://user-images.githubusercontent.com/23108066/146822787-af93cf1f-1c1d-46d0-bb68-5ed2df0bc3d5.png)
This is currently the turtle api, which uses a block position as well which is wrapped in the bukkit location. I'm not sure it's really worth the inconvenience it brings + inconsistency here, especially since really this represents a turtle that has an active world. If the entity isn't in a world, it'll just be null.